<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Introduction</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>

<!-- MathJax scripts -->
<script type="text/javascript" src="https://cdn.bootcss.com/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML">
</script>


<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<pre><code class="r">library(&quot;knitr&quot;)
require(&quot;Ravages&quot;)
</code></pre>

<h2>Introduction</h2>

<p>Ravages was developped to simulate genetic data and to perform burden tests (a type of rare variant association tests) on more than two groups of individuals. Ravages relies on the package Gaston developped by Herve Perdry and Claire Dandine-Roulland. Most functions are written in C++ thanks to the packages Rcpp, RcppParallel and RcppEigen.<br/>
Functions of Ravages use bed.matrix to manipulate genetic data as in the package Gaston (see documentation of this package for more details).<br/>
In this vignette, we show how to simulate genetic data and we illustrate association tests using these simulated data.
To learn more about all options of the functions, the reader is advised to look at the manual pages.</p>

<p>##Defining genomic regions
For rare variant association tests, the unit of analysis is not a single variant but a genomic region, typically a gene. The difficulty in this type of study is therefore to define the genomic region. The function <strong>set.genomic.region()</strong> can be used to group variant according to given gene positions. It works on a bed.matrix and simply adds a column &ldquo;genomic.region&rdquo; to the slot x@snps containing the gene assigned to each variant. By default, only variants within known genes will be assigned to a genomic region, the other ones being left out. Gene positions can be extended to annotate more variants using the argument <em>flank.width</em> corresponding to the number of base pair upstream and downstream the gene. If <em>flank.width=Inf</em>, each variant will be assigned to the nearest gene. If two genes overlap, variants in the overlapping zone will be attributed to the second one (in the order given by the position of their starting point on the genome).<br/>
The files <strong>genes.b37</strong> and <strong>genes.b38</strong> available in Ravages which contain gene positions from ENSEMBL versions GRCh37 and GRCh38 can be used to define gene positions. An example of this function is shown below using the example data &ldquo;LCT&rdquo; from gaston.</p>

<pre><code class="r">x &lt;- as.bed.matrix(LCT.gen, LCT.fam, LCT.bim)
#Group variants within known genes
x &lt;- set.genomic.region(x)
table(x@snps$genomic.region)
</code></pre>

<pre><code>## 
## R3HDM1  UBXN4    LCT   MCM6   DARS 
##    165     91    128     67     49
</code></pre>

<pre><code class="r">#Group variants within know genes by extended their positions
#500bp upstream and downstream
x &lt;- set.genomic.region(x, flank.width=500)
table(x@snps$genomic.region)
</code></pre>

<pre><code>## 
## R3HDM1  UBXN4    LCT   MCM6   DARS 
##    165     93    130     69     49
</code></pre>

<p>##Simulation of genetic data
Genetic data for two or more groups of individuals can be simulated using the package Ravages.</p>

<p>###Calculation of genotypic frequencies in each group of individuals
The first step to simulate genetic data is to compute genotypic frequencies in each group of individuals based on frequencies in the general population and on genetic relative risk (GRR) values. GRR correspond to the increased risk of a disease for a given genotype compared to the homozygous reference genotype. More precisely, the GRR associated to the heterozygous genotype in the group \(c\) corresponds to the ratio between the penetrance of phenotype \(c\) for the heterozygous genotype and the penetrance of phenotype \(c\) for the homozygous reference genotype as follow: \[GRR_{Aa} = \frac{P(Y=c | Aa)}{P(Y=c | AA)}\] With A the reference allele, and a the alternate allele. The frequency of each genotype in each group of cases \(c\) can be calculated using Bayes theorem:
\[P(Aa | Y=c) = \frac{P(Y=c | Aa)*P(Aa)}{\sum_{Geno=AA, Aa, aa} P(Y=c | Geno)*P(Geno)} = \frac{GRR_{Aa}*P(Aa)}{P(AA)+GRR_{Aa}*P(Aa)+GRR_{aa}*P(aa)}\]
P(AA), P(Aa) and P(aa) correspond to the genotypic probabilities in the general population.
The three genotypic frequencies can then be calculated in the controls group using the rule of total probability:
\[P(Geno | Y=1)=P(Geno) - \sum{_{c=2}^{c=C} P(Geno | Y=c)*P(Y=c)}\]
The function <strong>genotypic.freq()</strong> performs these calculations to obtain the three genotypic frequencies in the different groups of individuals. To do so, the user needs to give <em>P(Y=c)</em>, the prevalence of each group of cases (argument <em>baseline</em>); and the GRR values. GRR values need to be in a matrix with one row per cases group and one column per variant. If there is no supposed link between the GRR associated to the heterozygous genotype and the GRR associated to the homozygous alternate genotype (general model of the disease, \(genetic.model=&ldquo;general&rdquo;\)), the user needs to specify two GRR matrices: one for \(GRR_{Aa}\) (argument <em>GRR.het</em>) and one for \(GRR_{aa}\) (argument <em>GRR.homo.alt</em>). If <em>genetic.model=&ldquo;recessive&rdquo;</em>, <em>&ldquo;multiplicative&rdquo;</em> or <em>&ldquo;dominant&rdquo;</em>, only one GRR matrix is needed. By default, <em>genetic.model=&ldquo;multiplicative&rdquo;</em> and thus the GRR associated to the homozygote alternate genotype is the square of the GRR associated to the heterozygous genitype. <strong>genotypic.freq()</strong> will return a list with three matrices, one for each genotype containing the genotypic frequencies, with one row per group of individuals and one column per variant.<br/>
To help the user with the construction of the GRR matrix for <strong>genotypic.freq()</strong>, we implemented the function <strong>GRR.matrix()</strong>.<br/>
To use this function, the user needs to specify how the GRR should be calculated (argument <em>GRR</em>):  </p>

<ul>
<li>the user can choose to give the same GRR to all the variants (<em>GRR=&ldquo;constant&rdquo;</em>), its value being specified to the argument <em>GRR.value</em>;<br/></li>
<li>it is also possible to compute the GRR by using the formula from the publication presenting the method SKAT (<em>GRR=&ldquo;SKAT&rdquo;</em>);<br/></li>
<li>finally, the user can choose to calculate the GRR with its own function depending on MAF in the general population (<em>GRR=&ldquo;variable&rdquo;</em>), this function being specified to the argument <em>GRR.function</em>.<br/></li>
</ul>

<p>In the two last situations, a file containing the MAF in the general population with at least a column &ldquo;maf&rdquo; and a column &ldquo;gene&rdquo; should be given to the argument <em>genes.maf</em>. Two such files are available in Ravages: the file <em>Kryukov</em> containing MAF simulated under the demographic model of Kryukov and the file <em>GnomADgenes</em> containing MAF from the population NFE in GnomAD. As these files contain MAF for multiple genes, the user needs to specify which gene to choose to simulate the data with the argument <em>select.gene</em>. If this argument is empty, only the first gene will be kept in the analysis.<br/>
Finally, the multiplicative factor of the GRR between each group of cases compared to the first group of cases needs to be specified to the argument <em>GRR.multiplicative.factor</em> (number of values: number of cases groups - 1).<br/>
<strong>GRR.matrix()</strong> will return a GRR matrix in the appropriate format for the function <strong>genotypic.freq()</strong>.
Examples of these two functions are shown below:</p>

<pre><code class="r">#GRR calculated using the formula from the paper presenting SKAT,
#with values in the second group of cases twice as high as the first one

GRR.del &lt;- GRR.matrix(GRR = &quot;SKAT&quot;, genes.maf = Kryukov, n.case.groups = 2,
                      GRR.multiplicative.factor=2, select.gene = &quot;R1&quot;)
GRR.del[,1:5]
</code></pre>

<pre><code>##           [,1]      [,2]     [,3]      [,4]      [,5]
## [1,]  5.037728  6.222656 12.34472  9.042877  8.133663
## [2,] 10.075455 12.445313 24.68944 18.085755 16.267327
</code></pre>

<pre><code class="r">#Calculation of genotype frequencies in the two groups of cases and the controls group
#The previous GRR matrix is used with a multiplicative model of the disease
#All variants are deleterious and the prevalence in each group of cases is 0.001

geno.freq.groups &lt;- genotypic.freq(genes.maf = Kryukov, select.gene=&quot;R1&quot;, 
                                   GRR.het = GRR.del, baseline = c(0.001, 0.001),
                                   genetic.model = &quot;multiplicative&quot;)
str(geno.freq.groups)
</code></pre>

<pre><code>## List of 3
##  $ freq.homo.ref: num [1:3, 1:383] 1 0.999 0.998 1 1 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:3] &quot;controls&quot; &quot;cases_1&quot; &quot;cases_2&quot;
##   .. ..$ : NULL
##  $ freq.het     : num [1:3, 1:383] 1.87e-04 9.56e-04 1.91e-03 5.57e-05 3.52e-04 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:3] &quot;controls&quot; &quot;cases_1&quot; &quot;cases_2&quot;
##   .. ..$ : NULL
##  $ freq.homo.alt: num [1:3, 1:383] 7.90e-09 2.29e-07 9.15e-07 6.49e-10 3.11e-08 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:3] &quot;controls&quot; &quot;cases_1&quot; &quot;cases_2&quot;
##   .. ..$ : NULL
</code></pre>

<pre><code class="r">geno.freq.groups$freq.homo.alt[,1:5]
</code></pre>

<pre><code>##                  [,1]         [,2]         [,3]         [,4]         [,5]
## controls 7.897334e-09 6.485888e-10 7.480447e-14 6.568600e-12 2.503543e-11
## cases_1  2.288672e-07 3.106821e-08 4.778956e-11 9.067311e-10 2.469545e-09
## cases_2  9.145933e-07 1.242291e-07 1.911556e-10 3.626706e-09 9.877199e-09
</code></pre>

<p>It is also possible to calculate the MAF in each group of individuals as follow:</p>

<pre><code class="r">#MAF calculation for the five first variants
geno.freq.groups$freq.homo.alt[,1:5] + 0.5*geno.freq.groups$freq.het[,1:5]
</code></pre>

<pre><code>##                  [,1]         [,2]         [,3]         [,4]         [,5]
## controls 9.375276e-05 2.785699e-05 5.403418e-07 3.246158e-06 5.972867e-06
## cases_1  4.784006e-04 1.762618e-04 6.912999e-06 3.011198e-05 4.969452e-05
## cases_2  9.563437e-04 3.524614e-04 1.382590e-05 6.022214e-05 9.938410e-05
</code></pre>

<p>###Simulation of genotypes
In addition to calculate the genotypic frequencies in each group of individuals, it is possible to directly simulate these genotypes. This can be done using the function <strong>random.bed.matrix()</strong> which relies on the function <strong>genotypic.freq()</strong> explained previously.
The arguments <em>genes.maf</em>, <em>select.gene</em>, <em>baseline</em> and <em>genetic.model</em> are the same as in the function <strong>genotypic.freq()</strong>.<br/>
In <strong>random.bed.matrix()</strong>, the proportion of deleterious and protective variants simulated in the genomic region should be specified to <em>prop.del</em> and <em>prop.pro</em> respectively.<br/>
The argument <em>GRR.matrix.del</em> should contain a matrix with GRR values as if all the variants were deleterious (or protective). If <em>genetic.model=&ldquo;general&rdquo;</em>, two GRR matrices need to be given as a list to the argument <em>GRR.matrix.del</em> (one for the heterozygous genotype and the other for the homozygous alternate genotype).<br/>
If the user wants to simulate protective variants in addition to deleterious variants, a similar argument to <em>GRR.matrix.del</em> should be given to <em>GRR.matrix.pro</em> with GRR values as if all variants were protective. If the argument <em>GRR.matrix.pro</em> is empty and <em>prop.pro</em>&gt;0, the GRR values for protective variants will be calculated as 1/GRR of the deleterious variants.<br/>
These protective and deleterious GRR values will then be assigned to the sampled protective and deleterious variants in the simulations, the non-causal variants having GRR values of 1.<br/>
The size of the different groups of individuals should be a vector specified to <em>size</em>, and the user should choose whether the causal variants will be the same between the different groups of cases with the argument <em>same.variant</em>.<br/>
Using the argument <em>fixed.variant.prop</em>, the user needs also to choose if the argument <em>prop.del</em> (or <em>prop.pro</em>) corresponds to the final proportion of deleterious (or protective) variants, i.e. <em>fixed.variant.prop=TRUE</em> or to the probability associated to each variant of being deletrious (or protective), i.e. <em>fixed.variant.prop=FALSE</em>.<br/>
Finally, the number of genomic regions simulated is specified with the argument <em>replicates</em>.<br/>
<strong>random.bed.matrix()</strong> will return a bed matrix with the group of each individual in the field <em>@ped$pheno</em>, the first one being considered by default as the controls group, and the replicate number corresponding to the genomic region in the field <em>@snps$genomic.region</em>.<br/>
The example below shows how to simulate a group of 1,000 controls and two groups of 500 cases with 50% of deleterious variants having GRR values from the previous example. The deleterious variants are different between the two groups of cases and 5 genomic regions are simulated.</p>

<pre><code class="r">x &lt;- random.bed.matrix(genes.maf = Kryukov, size = c(1000, 500, 500), 
                       baseline = c(0.001, 0.001), GRR.matrix.del = GRR.del, 
                       prop.del = 0.5, prop.pro = 0, same.variant = FALSE,
                       fixed.variant.prop = TRUE, replicates = 5,
                       genetic.model = &quot;multiplicative&quot;, select.gene = &quot;R1&quot;)
x
</code></pre>

<pre><code>## A bed.matrix with 2000 individuals and 1915 markers.
## snps stats are set
##   There are  1636  monomorphic SNPs
## ped stats are set
</code></pre>

<pre><code class="r">table(x@ped$pheno)
</code></pre>

<pre><code>## 
##    0    1    2 
## 1000  500  500
</code></pre>

<pre><code class="r">table(x@snps$genomic.region)
</code></pre>

<pre><code>## 
##  R1  R2  R3  R4  R5 
## 383 383 383 383 383
</code></pre>

<p>##Rare variant definition
To perform rare variant analysis, it is important to define what is a rare variant in order to leave out common ones. The function <strong>filter.rare.variants()</strong> enables to keep only variants of interest based on a given MAF threshold. This function uses and returns a bed.matrix which can be filtered in three different ways:  </p>

<ul>
<li>If <em>filter=&ldquo;whole&rdquo;</em>, all the variants with a MAF lower than the threshold in the entire sample will be kept. </li>
<li>If <em>filter=&ldquo;controls&rdquo;</em>, all the variants with a MAF lower than the threshold in the controls group will be kept. In this situation, the controls group needs to be specified to the argument <em>ref.level</em>.</li>
<li>If <em>filter=&ldquo;any&rdquo;</em>, all the variants with a MAF lower than the threshold in any of the groups will be kept.<br/></li>
</ul>

<p>Monomorphic variants are also filtered out using this function. It is also possible to specify the minimum number of variants needed in a genomic region to keep it using the parameter <em>min.nb.snps</em>.</p>

<pre><code class="r">#Filter of rare variants to keep only variants with a MAF below 1% in any of the groups,
#and genomic regions with at least 5 variants
x.filter &lt;- filter.rare.variants(x, filter = &quot;any&quot;, maf.threshold = 0.01,
                                 min.nb.snps = 5)
x.filter
</code></pre>

<pre><code>## A bed.matrix with 2000 individuals and 274 markers.
## snps stats are set
## ped stats are set
</code></pre>

<pre><code class="r">table(x.filter@snps$genomic.region)
</code></pre>

<pre><code>## 
## R1 R2 R3 R4 R5 
## 56 56 63 57 42
</code></pre>

<p>##Rare variant association tests
We have implemented two rare variant burden association tests extensions: CAST and WSS. The general idea of burden tests is to compute a genetic score per individual and per gernomic region and and to test if it differs between the different groups of individuals. To extend these tests to more than two groups of individuals, a non-ordinal multinomial regression is used. The independant variable in this regression is the genetic effect of the gene represented by the genetic score and potential covariates can be added in the model. In addition to the genetic scores CAST and WSS directly implemented in the package, the user can specify another genetic score for the regression.</p>

<p>###Genetic score</p>

<p>####CAST
CAST is based on a binary score which has a value of one if an individual carries at least one variant in the considered genomic region, and 0 otherwise. A MAF threshold for the definition of a rare variant is therefore needed as an argument to <em>maf.threshold</em>. This score can be computed using the function <strong>CAST()</strong> as shown here on the data simulated previously:</p>

<pre><code class="r">#Calculation of the genetic score with a maf threshold of 1%
CAST.score &lt;- CAST(x = x.filter, genomic.region = x.filter@snps$genomic.region, 
                   maf.threshold = 0.01)
head(CAST.score)
</code></pre>

<pre><code>##       R1 R2 R3 R4 R5
## A0001  0  0  0  0  0
## A0002  0  0  0  0  0
## A0003  0  0  0  0  0
## A0004  0  0  0  0  0
## A0005  0  0  1  0  0
## A0006  0  0  0  0  0
</code></pre>

<p>No statistical test is performed using this function but the user can calculate the classical chi-square associated to the CAST score as follow:</p>

<pre><code class="r">#Chi-square test associated to the CAST score:
apply(CAST.score, 2, function (z) chisq.test(z, simulate.p.value = TRUE)$p.value)
</code></pre>

<pre><code>## R1 R2 R3 R4 R5 
##  1  1  1  1  1
</code></pre>

<p>####WSS
WSS (Weighted Sum Statistic) is based on a continuous score giving the highest weights to the rarest variants as follow: 
\[WSS_j = \sum_{i=1}^{R} I_{ij} * w_{i}\] with \[w_{i} = \frac{1}{\sqrt(t_{i} * q_{i} * 1-q{i})}\] and \[q_{i} = \frac{n_{i} + 1}{2*t_{i}+1}\]
Where \(n_{i}\) is the total number of minor alleles genotyped for variant \(i\), \(t_{i}\) is the total number of alleles genotyped for variant \(i\) and \(I_{ij}\) is the number of minor alleles of variant \(i\) for the invidual \(j\). In the original method, each variant is weighted according to its frequency in the controls group. In our version of WSS, the weights depend on allele frequency calculated on the entire sample. The function <strong>WSS()</strong> can be used to compute the WSS score as shown on the data simulated previously:</p>

<pre><code class="r">WSS.score &lt;- WSS(x = x.filter, genomic.region = x.filter@snps$genomic.region)
head(WSS.score)
</code></pre>

<pre><code>##       R1 R2       R3 R4 R5
## A0001  0  0 0.000000  0  0
## A0002  0  0 0.000000  0  0
## A0003  0  0 0.000000  0  0
## A0004  0  0 0.000000  0  0
## A0005  0  0 0.817007  0  0
## A0006  0  0 0.000000  0  0
</code></pre>

<h3>Regressions</h3>

<p>We have extended CAST and WSS using non-ordinal multinomial regression models. Let consider \(C\) groups of individuals including a group of controls (\(c=1\)) and \(C-1\) groups of cases with different sub-phenotypes of the disease. We can compute \(C-1\) probability ratios, one for each group of cases: \[ln \frac{P(Y_{j}=c)}{P(Y_{j}=1)} = \beta_{0,c} + \beta_{G,c}X_{G} + \beta_{k1,c}K_{1}+&hellip;+\beta_{kl,c}K_{l}\]
Where \(Y_{j}\) corresponds to the phenotype of the individual \(j\) and \(K_{l}\) is a vector for the $l$th covariate with the corresponding coefficient \(\beta_{kl}\). The genetic effect is represented by \(X_{G}\) and correspond to the genetic score CAST or WSS with \(\beta_{G,c}\) the log-odds ratio associated to this burden score.<br/>
The p-value associated to the genetic effect is calculated using a likelihood ratio test comparing this model to the same model without the genetic effect (null hypothesis). If only two groups are compared, a classical logistic regression is performed.<br/>
This regression can be performed on a bed.matrix using the function <strong>burden.mlogit()</strong> which relies on the package mlogit. To do so, the user needs to specify a vector with the phenotype of each individual (argument <em>group</em>) and the gene of each variant (argument <em>genomic.region</em>).<br/>
The  CAST or WSS genetic scores can be directly calculated in the regression (<em>burden=&ldquo;CAST&rdquo;</em> or <em>burden=&ldquo;WSS&rdquo;</em>). The user can also use another genetic score in the regression, which has to be specified as a matrix with one individual per row and one genomic region per column to <em>burden</em>.<br/>
The reference group of individuals should be given to the argument <em>ref.level</em>.
Potential covariates could also be included in the regression as a matrix with one row per individual and one column per covariate to the argument <em>data</em>. If only a subset of covariates from <em>data</em> are to be included in the model, a R formula should be given to <em>formula</em> with these covariates, otherwise all the covariates will be included.<br/>
<strong>burden.mlogit()</strong> will return the p-value associated to the regression for each genomic region. If there is a convergence problem with the regression, the function will return 1 in the column <em>is.err</em>. The odds ratio associated to each group of cases compared to the reference group with its confidence interval at a given alpha threshold (argument <em>alpha</em>) can also be obtained if <em>get.OR.value=TRUE</em>.<br/>
An example of the p-value and OR calculation with its 95% confidence interval for CAST and WSS on the data simulated previously using a non-ordinal multinomial regression on three groups is shown below.</p>

<pre><code class="r">##CAST
burden.mlogit(x=x.filter, group=x.filter@ped$pheno, 
              genomic.region=x.filter@snps$genomic.region,
              burden=&quot;CAST&quot;, maf.threshold=0.01, ref.level=&quot;0&quot;,
              alpha=0.05, get.OR.value=TRUE)
</code></pre>

<pre><code>##         p.value is.err     OR.1     OR.2 l.lower.1 l.lower.2 l.upper.1
## R1 8.933423e-06      0 2.673317 2.537870  1.681960  1.589350  4.248986
## R2 4.903604e-14      0 2.837530 6.292650  1.619144  3.819478  4.972735
## R3 1.191076e-17      0 3.833703 5.885735  2.416263  3.801971  6.082650
## R4 4.476728e-09      0 2.578948 4.964757  1.418031  2.902991  4.690288
## R5 4.653932e-10      0 2.941131 4.021918  1.829046  2.560373  4.729377
##    l.upper.2
## R1  4.052465
## R2 10.367240
## R3  9.111556
## R4  8.490834
## R5  6.317760
</code></pre>

<pre><code class="r">##WSS
burden.mlogit(x=x.filter, group=x.filter@ped$pheno, 
              genomic.region=x.filter@snps$genomic.region,
              burden=&quot;WSS&quot;, maf.threshold=0.01, ref.level=&quot;0&quot;,
              alpha=0.05, get.OR.value=TRUE)
</code></pre>

<pre><code>##         p.value is.err     OR.1      OR.2 l.lower.1 l.lower.2 l.upper.1
## R1 1.001287e-07      0 3.585691  5.601074  1.804395  2.931208  7.125482
## R2 1.523586e-15      0 5.489216 13.670886  2.472524  6.531020 12.186533
## R3 1.553276e-15      0 7.311996 13.217540  3.386063  6.327049 15.789805
## R4 2.834237e-10      0 3.633855  7.690030  1.770249  3.987000  7.459345
## R5 3.815630e-11      0 8.553566 12.597796  3.467277  5.241811 21.101139
##    l.upper.2
## R1  10.70276
## R2  28.61622
## R3  27.61214
## R4  14.83235
## R5  30.27665
</code></pre>

<pre><code class="r">##Simulation of covariates
covar &lt;- data.frame( sex = c(sample(0:1, 1000, TRUE, c(0.2,0.8)), 
                             sample(0:1, 1000, TRUE, c(0.8,0.2))), 
                     u = runif(1000))

##Regression with the covariate &quot;sex&quot;
burden.mlogit(x=x.filter, group=x.filter@ped$pheno, 
              genomic.region=x.filter@snps$genomic.region,
              burden=&quot;WSS&quot;, maf.threshold=0.01, ref.level=&quot;0&quot;,
              alpha=0.05, get.OR.value=TRUE,
              data=covar, formula = ~ sex)
</code></pre>

<pre><code>##         p.value is.err     OR.1      OR.2 l.lower.1 l.lower.2 l.upper.1
## R1 8.190158e-04      0 2.642542  4.061800  1.209084  1.905038  5.775471
## R2 2.339749e-10      0 4.808315 11.972409  2.005189  5.229189 11.530034
## R3 2.639033e-11      0 7.274889 13.317216  3.180923  5.950633 16.637943
## R4 2.168292e-06      0 2.887383  6.224677  1.320122  2.993379  6.315311
## R5 7.571158e-07      0 7.685355 11.185244  2.761780  4.069386 21.386456
##    l.upper.2
## R1  8.660307
## R2 27.411245
## R3 29.803253
## R4 12.944099
## R5 30.744116
</code></pre>

<p>##Data managment
Data in the plink format or in the vcf format can be loaded in R using the functions <strong>read.bed.matrix()</strong> and <strong>read.vcf()</strong> respectively from the package gaston.<br/>
If the data for the controls and the different groups of cases are in different files, they can be loaded separately and then combined using the function <strong>gaston:::rbind()</strong> as long as the same variants are present between the different groups of individuals.<br/>
An example is given below where the simulated data have been split according the the group of each individual, and then combined in a bed.matrix:</p>

<pre><code class="r">#Selection of each group of individuals
x.controls &lt;- select.inds(x.filter, pheno==&quot;0&quot;)
x.controls
</code></pre>

<pre><code>## A bed.matrix with 1000 individuals and 274 markers.
## snps stats are set
##   There are  194  monomorphic SNPs
## ped stats are set
</code></pre>

<pre><code class="r">x.cases1 &lt;- select.inds(x.filter, pheno==&quot;1&quot;)
x.cases1
</code></pre>

<pre><code>## A bed.matrix with 500 individuals and 274 markers.
## snps stats are set
##   There are  164  monomorphic SNPs
## ped stats are set
</code></pre>

<pre><code class="r">x.cases2 &lt;- select.inds(x.filter, pheno==&quot;2&quot;)
x.cases2
</code></pre>

<pre><code>## A bed.matrix with 500 individuals and 274 markers.
## snps stats are set
##   There are  115  monomorphic SNPs
## ped stats are set
</code></pre>

<pre><code class="r">#Combine in one file:
x.filter
</code></pre>

<pre><code>## A bed.matrix with 2000 individuals and 274 markers.
## snps stats are set
## ped stats are set
</code></pre>

<pre><code class="r">x.all &lt;- rbind(x.controls, x.cases1, x.cases2)
x.all
</code></pre>

<pre><code>## A bed.matrix with 2000 individuals and 274 markers.
## snps stats are set
## ped stats are set
</code></pre>

<p>##Power calculation
The power of burden tests extensions can be directly calculated on simulations previously explained using the function <strong>power()</strong>. Many arguments are needed to run this function, corresponding to the arguments from the simulations and the arguments from the regression models. The arguments needed to run the simulations should be given as a list to <em>model.pars</em>: this list should contain the arguments needed to <strong>random.bed.matrix</strong>; in addition, a GRR matrix should be given to <em>model.pars$GRR.matrix.del</em> (and/or <em>model.pars$GRR.matrix.pro</em>) or the arguments from the function <strong>GRR.matrix()</strong> should be given to compute this GRR matrix.<br/>
The significance threshold used to calculate the power should also be specified in the argument <em>alpha</em>. As this function calculates only the power of the tests, individual p-value and OR values are not returned.<br/>
An example of this function to estimate the power at a 0.1% significance threshold of CAST and WSS with 100 replicates using the same parameters for the simulations as previously is shown below:</p>

<pre><code class="r">power(alpha = 0.001, WSS = TRUE, CAST = TRUE, maf.threshold=0.01,
      filter = &quot;any&quot;, min.nb.snps = 5,
      pooled.analysis = TRUE, non.pooled.analysis = TRUE,
      analysis.by.group = TRUE, ref.level=&quot;0&quot;,
      formula=NULL, data=NULL,
      model.pars=list(genes.maf=Kryukov, select.gene=&quot;R1&quot;, size=c(1000, 500, 500),
                      baseline=c(0.001, 0.001), n.case.groups=2, replicates=100,
                      GRR=&quot;SKAT&quot;, GRR.multiplicative.factor=2,
                      same.variant=FALSE, fixed.variant.prop = TRUE,
                      genetic.model=&quot;multiplicative&quot;,
                      prop.del=0.5, prop.pro=0) )
</code></pre>

<pre><code>##                              power          se nb.replicates
## CAST                          1.00 0.000000000           100
## Cases1vsControls.CAST         0.65 0.047696960           100
## Cases2vsControls.CAST         0.99 0.009949874           100
## pooled.CAST                   0.99 0.009949874           100
## WSS                           1.00 0.000000000           100
## Cases1vsControls.WSS          0.68 0.046647615           100
## Cases2vsControls.WSS          0.99 0.009949874           100
## pooled.WSS                    1.00 0.000000000           100
## other.score                     NA          NA            NA
## Cases1vsControls.other.score    NA          NA            NA
## Cases2vsControls.other.score    NA          NA            NA
## pooled.other.score              NA          NA            NA
</code></pre>

<p>Another example below replicates one of the results from the paper &ldquo;Rare variant association testing for multi-category phenotype&rdquo; submitted to Genetic Epidemiology. These simulations correspond to the scenario DG where genetic effects are present only in the second group of cases with 50% of causal variants:</p>

<pre><code class="r">#GRR matrix used for the DG scenario where genetic effects are present only in 
#the second group of cases (i.e. GRR=1 in the first group of cases for all the variants)
GRR.DG &lt;- matrix(c(rep(1, nrow(subset(Kryukov, gene==&quot;R1&quot;))), 
                   exp(0.402 * abs(log10(subset(Kryukov, gene==&quot;R1&quot;)$maf)))),
                 nrow=2, byrow=TRUE)
GRR.DG[,1:5]
</code></pre>

<pre><code>##          [,1]     [,2]     [,3]     [,4]     [,5]
## [1,] 1.000000 1.000000  1.00000 1.000000 1.000000
## [2,] 5.037728 6.222656 12.34472 9.042877 8.133663
</code></pre>

<pre><code class="r">#Calculation of power
power(alpha=2.5e-6, filter=&quot;any&quot;, CAST=TRUE, WSS=TRUE, pooled.analysis=TRUE, 
      non.pooled.analysis=TRUE, analysis.by.group=TRUE, ref.level=&quot;0&quot;, 
      model.pars=list(genes.maf = Kryukov, select.gene = &quot;R1&quot;, size = c(1000, 500, 500), 
                      baseline = c(0.001, 0.001), n.case.groups = 2, replicates = 1000, 
                      genetic.model = &quot;multiplicative&quot;,prop.del = 0.5, prop.pro = 0, 
                      GRR.matrix.del=GRR.DG))
</code></pre>

<pre><code>##                              power          se nb.replicates
## CAST                         0.227 0.013246547          1000
## Cases1vsControls.CAST        0.000 0.000000000          1000
## Cases2vsControls.CAST        0.193 0.012480024          1000
## pooled.CAST                  0.020 0.004427189          1000
## WSS                          0.225 0.013205113          1000
## Cases1vsControls.WSS         0.000 0.000000000          1000
## Cases2vsControls.WSS         0.201 0.012672766          1000
## pooled.WSS                   0.021 0.004534203          1000
## other.score                     NA          NA            NA
## Cases1vsControls.other.score    NA          NA            NA
## Cases2vsControls.other.score    NA          NA            NA
## pooled.other.score              NA          NA            NA
</code></pre>

</body>

</html>
